### PR TITLE
Added support systemd for qt-service

### PR DESCRIPTION
**Pull request by EndrII**
Current qt-service do not working with systemd system, this is PR fix this problem.

I also added the ability to manually set the path to the executable file of the service, since sometimes a service needs a shell that configures its environment.

* setServiceExecutable

### CODE REVIEW

File: qtservice/src/qtservice_unix.cpp [hunk 1]
```diff
@@ -58,6 +58,43 @@
...
+#include <QFile>
+
+static bool isServiceInited = false;
+
+static bool isSystemD()
+{
+    return QFileInfo::exists("/etc/systemd/");
+}
+
+static QString systemDPass()
```
Comment (KangLin) [inline qtservice/src/qtservice_unix.cpp:70]: Is it better to name with systemDPath？

**Review by KangLin — CHANGES_REQUESTED** [review]

File: qtservice/src/qtservice_unix.cpp (hunk 1 above)
Comment (EndrII) [reply qtservice/src/qtservice_unix.cpp:70]: maybe you right

File: qtservice/src/qtservice_unix.cpp [hunk 2]
```diff
@@ -58,6 +58,43 @@
 #include <QMap>
 #include <QSettings>
 #include <QProcess>
+#include <QFile>
+
+static bool isServiceInited = false;
+
+static bool isSystemD()
+{
+    return QFileInfo::exists("/etc/systemd/");
```
Comment (EndrII) [inline qtservice/src/qtservice_unix.cpp:67]:
```suggestion
    return QFileInfo::exists(systemDPass());
```

Comment (KangLin) [reply qtservice/src/qtservice_unix.cpp:67]: static QString systemDPass() rename static QString systemDPath()

File: qtservice/src/qtservice_unix.cpp (hunk 1 above)
Comment (KangLin) [reply qtservice/src/qtservice_unix.cpp:70]: static QString systemDPass() rename static QString systemDPath()

**Review by KangLin — CHANGES_REQUESTED** [review]

File: qtservice/src/qtservice_unix.cpp (hunk 1 above)
Comment (EndrII) [inline qtservice/src/qtservice_unix.cpp:70]:
```suggestion
static QString systemDPath()
```

File: qtservice/src/qtservice_unix.cpp [hunk 3]
```diff
@@ -58,6 +58,43 @@
...
+}
+
+static void initService()
+{
+
+    if (isSystemD() && !isServiceInited)
+    {
+        QSettings::setPath(QSettings::NativeFormat,
+                           QSettings::SystemScope,
+                           systemDPass());
```
Comment (EndrII) [inline qtservice/src/qtservice_unix.cpp:82]:
```suggestion
                           systemDPath());
```

File: qtservice/src/qtservice_unix.cpp [hunk 4]
```diff
@@ -191,7 +231,23 @@ QString QtServiceController::serviceFilePath() const
 
 bool QtServiceController::uninstall()
 {
-    QSettings settings(QSettings::SystemScope, "QtSoftware");
+
+    if (isSystemD()) {
+        uninstallSysD(systemDPass() + serviceName() + ".service");
```
Comment (EndrII) [inline qtservice/src/qtservice_unix.cpp:236]:
```suggestion
        uninstallSysD(systemDPath() + serviceName() + ".service");
```

File: qtservice/src/qtservice_unix.cpp [hunk 5]
```diff
@@ -240,7 +296,12 @@ bool QtServiceController::sendCommand(int code)
 
 bool QtServiceController::isInstalled() const
 {
-    QSettings settings(QSettings::SystemScope, "QtSoftware");
+
+    if (isSystemD()) {
+        return QFileInfo::exists(systemDPass() + serviceName() + ".service");
```
Comment (EndrII) [inline qtservice/src/qtservice_unix.cpp:301]:
```suggestion
        return QFileInfo::exists(systemDPath() + serviceName() + ".service");
```

File: qtservice/src/qtservice_unix.cpp [hunk 6]
```diff
@@ -417,10 +478,62 @@ bool QtServiceBasePrivate::start()
...
+    settings.endGroup();
+
+    settings.beginGroup("Install");
+    settings.setValue("WantedBy", "multi-user.target");
+    settings.endGroup();
+
+    settings.sync();
+
+    bool renamed = QFile(systemDPass() + controller.serviceName() + ".conf").rename(
+                systemDPass() + controller.serviceName() + ".service");
```
Comment (EndrII) [inline qtservice/src/qtservice_unix.cpp:520]:
```suggestion
    bool renamed = QFile(systemDPath() + controller.serviceName() + ".conf").rename(
                systemDPath() + controller.serviceName() + ".service");
```

File: qtservice/src/qtservice_unix.cpp (hunk 1 above)
Comment (EndrII) [reply qtservice/src/qtservice_unix.cpp]: done

File: qtservice/src/qtservice_unix.cpp [hunk 7]
```diff
@@ -149,7 +189,7 @@ QString QtServiceBasePrivate::filePath() const
 
 QString QtServiceController::serviceDescription() const
 {
-    QSettings settings(QSettings::SystemScope, "QtSoftware");
+    auto &settings = getSettings(serviceName());
```
Comment (KangLin) [inline qtservice/src/qtservice_unix.cpp:192]: There are not modify.

File: qtservice/src/qtservice_unix.cpp [hunk 8]
```diff
@@ -58,6 +58,43 @@
...
+    if (isSystemD() && !isServiceInited)
+    {
+        QSettings::setPath(QSettings::NativeFormat,
+                           QSettings::SystemScope,
+                           systemDPath());
+        isServiceInited = true;
+    }
+}
+
+static QSettings& getSettings(const QString& serviceName)
```
Comment (KangLin) [inline qtservice/src/qtservice_unix.cpp:87]: The function isn't need.

File: qtservice/src/qtservice_unix.cpp [hunk 9]
```diff
@@ -58,6 +58,43 @@
...
+{
+    return "/etc/systemd/system/";
+}
+
+static void initService()
+{
+
+    if (isSystemD() && !isServiceInited)
+    {
+        QSettings::setPath(QSettings::NativeFormat,
```
Comment (KangLin) [inline qtservice/src/qtservice_unix.cpp:80]: There are not setPah. It is need caller set. Please see help of QSettings::setPath.
1. The caller is call QSettings::setPath
2. Call QSettings::setPath in Constructor

File: qtservice/src/qtservice_unix.cpp [hunk 10]
```diff
@@ -163,7 +203,7 @@ QString QtServiceController::serviceDescription() const
 
 QtServiceController::StartupType QtServiceController::startupType() const
 {
-    QSettings settings(QSettings::SystemScope, "QtSoftware");
+    auto &settings = getSettings(serviceName());
```
Comment (KangLin) [inline qtservice/src/qtservice_unix.cpp:206]: There are not modify.

File: qtservice/src/qtservice_unix.cpp [hunk 11]
```diff
@@ -177,7 +217,7 @@ QtServiceController::StartupType QtServiceController::startupType() const
 
 QString QtServiceController::serviceFilePath() const
 {
-    QSettings settings(QSettings::SystemScope, "QtSoftware");
+    auto &settings = getSettings(serviceName());
```
Comment (KangLin) [inline qtservice/src/qtservice_unix.cpp:220]: There are not modify.

File: qtservice/src/qtservice_unix.cpp [hunk 12]
```diff
@@ -191,7 +231,23 @@ QString QtServiceController::serviceFilePath() const
...
+
+bool QtServiceController::uninstallSysD(const QString &serviceFile)
+{
+    return QFile::remove(serviceFile);
+
+}
+
+bool QtServiceController::uninstallUpStart()
+{
+    auto &settings = getSettings(serviceName());
```
Comment (KangLin) [inline qtservice/src/qtservice_unix.cpp:250]: There are not modify. use QSettings settings(QSettings::SystemScope, "QtSoftware");

File: qtservice/src/qtservice_unix.cpp [hunk 13]
```diff
@@ -240,7 +296,12 @@ bool QtServiceController::sendCommand(int code)
 
 bool QtServiceController::isInstalled() const
 {
-    QSettings settings(QSettings::SystemScope, "QtSoftware");
+
+    if (isSystemD()) {
+        return QFileInfo::exists(systemDPath() + serviceName() + ".service");
+    }
+
+    auto &settings = getSettings(serviceName());
```
Comment (KangLin) [inline qtservice/src/qtservice_unix.cpp:304]: There are not modify. use QSettings settings(QSettings::SystemScope, "QtSoftware");

File: qtservice/src/qtservice_unix.cpp [hunk 14]
```diff
@@ -297,7 +358,7 @@ private slots:
 };
 
 QtServiceSysPrivate::QtServiceSysPrivate()
-    : QtUnixServerSocket(), ident(0), serviceFlags(0)
+    : QtUnixServerSocket(), ident(nullptr), serviceFlags(nullptr)
```
Comment (KangLin) [inline qtservice/src/qtservice_unix.cpp:361]: There are not modify.

File: qtservice/src/qtservice_unix.cpp [hunk 15]
```diff
@@ -417,10 +478,62 @@ bool QtServiceBasePrivate::start()
...
+    return (ret == QSettings::NoError);
+}
+
+bool QtServiceBasePrivate::installUpStart(const QString &account, const QString &password)
 {
     Q_UNUSED(account)
     Q_UNUSED(password)
-    QSettings settings(QSettings::SystemScope, "QtSoftware");
+
+    auto &settings = getSettings(controller.serviceName());
```
Comment (KangLin) [inline qtservice/src/qtservice_unix.cpp:536]: There are not modify.

**Review by KangLin — CHANGES_REQUESTED** [review]

**Comment by EndrII** [comment]
This is not actual anymore

**Pull request closed without merging**